### ISSUE TITLE
Image labeling at FOV level

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal#64]
Hi @gusqgm and @jluethi, let's discuss here the details of the image labeling task.

**Comment by tcompa** [comment]
As of b63c183560be76e699d9833fbe8a508c2afe618c, we have a preliminary working version that sets the correct scales and writes through `ome_zarr.writer`.

Segmentation is performed on a mid-res level (level 2), and scales are set as they should in `plate.zarr/B/03/0/labels/label_image/.zattrs`:
```
{
    "image-label": {
        "version": "0.4"
    },
    "multiscales": [
        {
            "axes": [
                {
                    "name": "z",
                    "type": "space"
                },
                {
                    "name": "y",
                    "type": "space"
                },
                {
                    "name": "x",
                    "type": "space"
                }
            ],
            "datasets": [
                {
                    "coordinateTransformations": [
                        {
                            "scale": [
                                1.0,
                                4.0,
                                4.0
                            ],
                            "type": "scale"
                        }
                    ],
                    "path": "0"
                },
                {
                    "coordinateTransformations": [
                        {
                            "scale": [
                                1.0,
                                8.0,
                                8.0
                            ],
                            "type": "scale"
                        }
                    ],
                    "path": "1"
                },
                {
                    "coordinateTransformations": [
                        {
                            "scale": [
                                1.0,
                                16.0,
                                16.0
                            ],
                            "type": "scale"
                        }
                    ],
                    "path": "2"
                },
                {
                    "coordinateTransformations": [
                        {
                            "scale": [
                                1.0,
                                32.0,
                                32.0
                            ],
                            "type": "scale"
                        }
                    ],
                    "path": "3"
                }
            ],
            "name": "label_image",
            "version": "0.4"
        }
    ]
}
```

Napari visualization confirms that the labels correspond to the image, also upon zooming in/out.

**Comment by tcompa** [comment]
The current choice (see code in https://github.com/fractal-analytics-platform/fractal/commit/b63c183560be76e699d9833fbe8a508c2afe618c) should be a bit more general than just applying the `1/coarsening^level` approach, for cases where we built a pyramid with (e.g.) coarsening factor 3, and we had to trim some parts of the array because of non-commensurability.

However, at the moment we are defining the scales for the labels by looking at the shapes of the images. Depending on how `ome_zarr` constructs the pyramids (e.g. with downsampling factor 3, often leading to non-commensurability), we may have to fix the current approach.

This could be a bit annoying because the pyramids are built *within* `write_labels`, but we can find several ways to fix it.

**Comment by tcompa** [comment]
Rapid GPU/Cellpose benchmark (still with a standalone python script):

An array with shape `(19, 2430, 2560)` was segmented in ~3 minutes.
An array with shape `(19, 4860, 5120)` was segmented in ... (to be updated)

**Comment by jluethi** [comment]
If the `(19, 4860, 5120)` case can be processed on our current cluster GPUs, then I'd say we're good to go forward to implement segmentation per original FOV (chunk at level 0 of the pyramid) at full resolution and we can worry about implementations at higher pyramid levels a bit later.

Priority order would then be (as just discussed in the call):
1. Segmentation at level 0 by site (original FOVs) in 2D & 3D
2. Segmentation for a full well, e.g. the full well in 2D (may still work at full resolution, i.e. pyramid level 0)
3. Segmentation at user-selected levels for 2D or 3D. e.g. Segment sites at pyramid level 1 or segment the whole well at level 2.

Let's make sure the first priority works and get this into Fractal. Then we can expand to priorities 2 & 3.

To enable case 3, we need to be able to get original FOVs in the well back. That would have been easy in a multi-FOV Zarr file. But I think we'll find good ways to do this in the single-FOV Zarr file as well. My current favorite path towards that: Save ROIs (region of interest, i.e. corner coordinates + sizes) in an OME-NGFF table. Then loop over all ROIs, extract the information of the relevant regions from the pyramidal chunks based on their position in a well and their scale parameter.

**Comment by tcompa** [comment]
> If the `(19, 4860, 5120)` case can be processed on our current cluster GPUs, then I'd say we're good to go forward to implement segmentation per original FOV (chunk at level 0 of the pyramid) at full resolution and we can worry about implementations at higher pyramid levels a bit later.

Apparently it is already too large (see edited comment above).

**Comment by jluethi** [comment]
We've also started discussing the relabeling issue:
When we run segmentation per chunk, each chunk has labels going from 1 to n. If the chunks are then combined again into a large image, the labels within the "well-image" are non-unique. If we want to map measurements to labels, we typically require unique labels per image though.

One solution to this is to "relabel" the different chunks to ensure label uniqueness. A description of the problem for dask and a relabeling strategy (though not for integer labels I think) can be found here: https://github.com/dask/dask/issues/2768

**Comment by jluethi** [comment]
> Apparently it is already too large (see edited comment above).

Hmm, so we are close to the memory limit. But if `(19, 2430, 2560)` (not just `(10, 2430, 2560)`) still passes, we should be good to go for a first implementation :)

**Comment by tcompa** [comment]
> But if `(19, 2430, 2560)` (not just `(10, 2430, 2560)`) still passes

It does pass.

**Comment by tcompa** [comment]
Possible issue: it looks like [`ome_zarr`'s `writer`](https://github.com/ome/ome-zarr-py/blob/master/ome_zarr/writer.py) only deals with numpy arrays, and not dask arrays. I think this means we won't be able to store labels for a single (say) 9x8 well with at least 10 Z planes.

Possible ideas:
1. Checking whether `writer` actually could be updated to support dask arrays. Warning: this wouldn't be a five-minutes fix, because `writer` is also connected to their other modules (`scale`, for instance). Plus we would first need green light from their devs, since it is quite an important change (essentially one should make all of `ome_zarr` dask-compatible, I guess).
3. We switch back to our own pyramid-creation utils.

I'll try 2 first, to see whether I can make it work quickly.

**Comment by tcompa** [comment]
Test code related to the previous comment:
```
import numpy as np
import shutil
import dask.array as da
import zarr

from ome_zarr.io import parse_url
from ome_zarr.writer import write_labels

store = parse_url("plate.zarr", mode="w").store
root = zarr.group(store=store)
label_name = "label_image"
label_axes = ["z", "y", "x"]  

mask = np.ones((10, 200, 200), dtype=np.uint16)
shutil.rmtree("plate.zarr")
write_labels(mask, group=root, name=label_name, axes=label_axes)
print("Successfully written the np array")

mask = da.ones((10, 200, 200), dtype=np.uint16).compute()
shutil.rmtree("plate.zarr")
write_labels(mask, group=root, name=label_name, axes=label_axes)
print("Successfully written the computed dask array")

mask = da.ones((10, 200, 200), dtype=np.uint16)
shutil.rmtree("plate.zarr")
write_labels(mask, group=root, name=label_name, axes=label_axes)
print("Successfully written the dask array")
```

**Comment by jluethi** [comment]
@tcompa Ah, makes sense! The whole OME-Zarr library isn't really optimized for out-of-core writing, which is why we have a lot of our own parsing to OME-Zarrs for the images as well. When we were still thinking about saving per original FOV, using the OME-Zarr library would have made sense.

But in the current framework of single-FOV wells, I think it makes more sense to create our own writer for labels now as well. Getting the OME-Zarr library to be out-of-core writing compatible is interesting, but a larger undertaking.

**Comment by tcompa** [comment]
A question about re-labeling:

Did I understand correctly that we could expect up to ~1000 labels in a single 3D column (say `10x2160x2560`)?
In that case, the total number of labels in a large well (say 72 sites or more) after re-labeling will be larger than what allowed by `uint16`, and we'll need to switch to `uint32`. Is that OK or should we discuss it further?

**Comment by jluethi** [comment]
For our current test cases, most wells contain ~50k cells, which would often still be below the limit. But there may be wells in the 23 well test case that actually go above this and other use cases where people have more than 65535 objects in a well.

If we can test `uint32` label images and they work with the current pipeline, that should avoid the issue for any reasonable use-cases we use with label images in a well.

I haven't worked with `uint32` label images in napari or ome-zarr before though, so don't yet know whether they would make any issues

**Comment by jluethi** [comment]
Size-wise, label images are typically very small, i.e. a few kbs for a chunk that is a few MBs, so even doubling bit depth shouldn't be an issue there. If you have a quick test to see that uint32 label images are still opened in napari, that would be the only check I think will be necessary.
Let me know if you have a fast way to test this, otherwise I can test this on some example data this afternoon.

**Comment by tcompa** [comment]
Questions keep coming:

We are not yet parsing microscope metadata, so for the moment I will set the `anisotropy` default to the one that matters for current test datasets. From https://user-images.githubusercontent.com/18033446/174331876-c80c780f-95da-4c38-af6e-affa9e8d9c69.png and
```
        anisotropy: float (optional, default None)
            for 3D segmentation, optional rescaling factor (e.g. set to 2.0 if Z is sampled half as dense as X or Y)
```
I would say that `anisotropy=5/0.325=15.38`.

Does this number sound reasonable?

**Comment by tcompa** [comment]
> If you have a quick test to see that uint32 label images are still opened in napari, that would be the only check I think will be necessary.

I opened some uint32 labels in napari (this is for the MIP of a 9x8 well, with ~40k labels in total) and everything seems to work.

![Screenshot from 2022-07-04 12-22-11](https://user-images.githubusercontent.com/3862206/177135811-36c144b1-a0bb-48b4-9089-bde765577e10.png)

**Comment by jluethi** [comment]
> I would say that anisotropy=5/0.325=15.38.

How do we get to the 5 there? For this experiment, Z steps are 1um, xy dimension at full resolution is 0.1625um. At level 1 (with 2x coarsening), it's 0.325

So if we work at level 1, 1/0.325 ~= 3 is a reasonable choice (which is why I also had 3 as my original default value). Would be good if that's something the user can overwrite if needed (sometimes models work better by not actually using the physical anisotropy exactly)

> I opened some uint32 labels in napari 

This looks awesome! `uint32` label images sound good to me in that case! :)

**Comment by tcompa** [comment]
> How do we get to the 5 there?

I got it from the `pixel_size_z` column in the linked table, but then it was probably something else.

> For this experiment, Z steps are 1um, xy dimension at full resolution is 0.1625um. At level 1 (with 2x coarsening), it's 0.325
> So if we work at level 1, 1/0.325 ~= 3 is a reasonable choice (which is why I also had 3 as my original default value).

Since I'm working at level 0, I should set `anisotropy~6`, right?

> Would be good if that's something the user can overwrite if needed (sometimes models work better by not actually using the physical anisotropy exactly)

Sure, that's the plan. More generally, we aim at very little hardcoded parameters (apart from very basic things like the image size).

**Comment by tcompa** [comment]
As of https://github.com/fractal-analytics-platform/fractal/commit/8e2998ad616941614962440002210617b39028da, there is a (very preliminary) version of Fractal integration. In principle it works (see `examples/example_uzh_1_well_2x2_sites.sh`), but there are ongoing discussions concerning scalability (both for memory and GPU resources), more updates as soon as possible.

I would not test this on the 23-wells yet, but probably one well with 9x8 sites should work.

In principle the task works both for 3D and 2D (MIP) data, but at the moment Fractal always runs on the 3D images.

**Comment by tcompa** [comment]
Since memory usage for the labeling task is going to be an important issue here, let's keep some reference info here.
A line which is quite memory-expensive, in cellpose, is https://github.com/MouseLand/cellpose/blob/main/cellpose/core.py#L599:
```
yf = np.zeros((3, self.nclasses, imgs.shape[0], imgs.shape[1], imgs.shape[2]), np.float32)
```

As a reference, this could look like this:
```
$ python
Python 3.9.12 (main, Apr  5 2022, 06:56:58) 
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np

>>> x = np.random.randint(0, 100000, size=(3, 3, 10, 2160, 2560)).astype(np.float32)
>>> print(f"dype={x.dtype}, size={x.size}, size*itemsize={x.size*x.itemsize/1000000} GB")
dype=float32, size=497664000, size*itemsize=1990.656 GB

>>> x = np.random.randint(0, 100000, size=(3, 3, 15, 2160, 2560)).astype(np.float32)
>>> print(f"dype={x.dtype}, size={x.size}, size*itemsize={x.size*x.itemsize/1000000} GB")
dype=float32, size=746496000, size*itemsize=2985.984 GB
```

**Comment by tcompa** [comment]
Quick update on labeling/parallelization/memory (probably best to discuss this tomorrow)

## TL;DR
1. We are currently running cellpose sequentially over sites, which is possibly slower but safer memory-wise;
2. We are currently storing the array of labels for a whole well, non-lazily. Memory-wise, this scales badly for larger wells, but at the moment we don't have a better way (due to a mix of relabeling needs and the previous point).

We can probably live with point 1, but point 2 doesn't look good. More work is needed.

## More details

We are testing a single well with 9x8 sites in a single-FOV scheme.
[_Note: the single-FOV approach is the only one that we are currently supporting, as of recent decisions, and things won't change unless we explicitly change our mind (that is, we are not running anything multi-FOV)._]

For labeling, at the moment we go through the 72 sites sequentially, i.e. there are never two cellpose calculations running at the same time.
Early experiments with more flexible approaches (letting dask organize the 72 parallel executions, through `map_blocks` or similar, including https://github.com/dask/dask/issues/2768) led to memory errors within cellpose (for instance in [the line mentioned above](https://github.com/MouseLand/cellpose/blob/main/cellpose/core.py#L599)), and our obvious guess is that this was due to simultaneous execution of more than one cellpose calculation.

At the moment, we are not able to generate the array of labels of a whole well in a lazy way, and that's the reason for the large CPU-memory usage. For a well of 72 sites with 19 Z planes, this array takes around 30 GB (for uint32 labels). Which is still doable on our current RAM (64 G), but not really scalable. A quick mitigation is to switch back to uint16 labels, and only use uint32 when the number of labels becomes larger than 65k. But we would be just pushing the problem a bit further, not solving it.

If we dropped relabeling (which is not an option, in our understanding), we could try to construct this array lazily, by just:
1. Execute cellpose for chunk 1/72 (first column of 3D data), and store it on disk;
2. Execute cellpose for chunk 2/72 (second column of 3D data), and store it on disk;
4. ...


This would still have two problems (1. concurrent execution of cellpose calculations can lead to memory errors, 2. lack of relabeling), but at least we would not be building the whole `9x8x19x2160x2560` array.

**Comment by tcompa** [comment]
Some thoughts as of yesterday's Fractal meeting:

1. A first useful update would be to separate labeling and relabeling.  
    * In this way, it would be simpler to make the labeling part lazy, with great benefits in terms of CPU memory.
    * Missing: we still need to limit the number of simultaneous Cellpose calls on the same GPU, but how to do it from dask is currently unclear. Possibly useful: https://github.com/dask/distributed/issues/800, https://github.com/rapidsai/cuml/issues/1822. If it's not easily doable, we can create small batches of (say) four sites and execute them together.
    * The relabeling is probably something which can be done naively (label files are not so large), and it should also be optional.
2. The next required use case is to run segmentation on a whole well, for 2D data (e.g. MIP). This will require:
    * A different structure of the task (re: what we parallelize over). See also point 3.
    * The possibility of working at lower-resolution levels (most likely).
 3. If/when we implement the ROI definition as in https://github.com/fractal-analytics-platform/fractal/issues/95, the task can be restructured in a way that includes both use cases (parallelized over sites, or directly on the whole well, at whatever resolution level), and opens the door to parallelization over arbitrary ROIs.

**Comment by tcompa** [comment]
> * Missing: we still need to limit the number of simultaneous Cellpose calls on the same GPU, but how to do it from dask is currently unclear.

Probably useful:
https://docs.dask.org/en/stable/scheduler-overview.html#configuring-the-schedulers

Here's a minimal working example on a CPU:
```
mport time
import numpy as np
import dask
import dask.array as da
from concurrent.futures import ThreadPoolExecutor


n_threads = 4

def heavy_cellpose_function(block, block_info=None):
    print(f"Now running heavy cellpose function for chunk {block_info[None]['chunk-location']}")
    time.sleep(10)
    return np.ones_like(block, dtype=np.uint16) * np.random.randint(10)
x = da.zeros((8, 12), chunks=(4, 3), dtype=np.uint16)

# Extract total number of chunks
n_chunks = (x.shape[0] // x.chunksize[0]) * (x.shape[1] // x.chunksize[1])
print(f"Number of chunks: {n_chunks}")
print(f"Number of threads: {n_threads}")
print()

t0 = time.perf_counter()
with dask.config.set(pool=ThreadPoolExecutor(n_threads)):
    y = x.map_blocks(heavy_cellpose_function, chunks=(4, 3), meta=np.array((), dtype=np.uint16))
    y.to_zarr("data.zarr")
t1 = time.perf_counter()
print()
print(f"Elapsed time: {t1 - t0}")
print()
```
with output:
```
Number of chunks: 8
Number of threads: 4

Now running heavy cellpose function for chunk (0, 0)
Now running heavy cellpose function for chunk (0, 1)
Now running heavy cellpose function for chunk (0, 2)
Now running heavy cellpose function for chunk (0, 3)
Now running heavy cellpose function for chunk (1, 0)
Now running heavy cellpose function for chunk (1, 1)
Now running heavy cellpose function for chunk (1, 2)
Now running heavy cellpose function for chunk (1, 3)

Elapsed time: 20.044721081001626
```

Now we should check whether it also works on the GPU.

**Comment by tcompa** [comment]
Still on "running cellpose for several sites at the same time on the GPU":

We consider sites of shape `(10,2160,2560)`, and we limit the number of concurrent cellpose executions as in https://github.com/fractal-analytics-platform/fractal/issues/64#issuecomment-1177229164. 

We run the test on a 2x2 well, but we only count the sites which are actually run in parallel (i.e. if we have 4=3+1 sites and run them in batches of 3, we only look at the timing for the first three and we ignore the last one). This is the only relevant measure when scaling to large wells.

| batch_size | time_per_site | time_per_site / batch_size | Speed-up factor |
| - | - | - | - |
| 1 | ~4 min | ~4 min| 1 |
|2 | ~5 min | ~2.5 min | ~1.6 |
| 3 | ~6 min |  ~2 min | ~2 |
| 4 | MemoryError | - | - |

This suggests that:
1. It is beneficial to add a little bit of same-GPU parallelism.
2. This wont' scale to anything more than small batches of 2-3 sites. And for larger datasets (i.e. more Z planes), the maximum allowed batch_size could even be one.

**Comment by jluethi** [comment]
Great summary of the discussion at yesterday's meeting @tcompa :)

> This suggests that:
> 
> 1. It is beneficial to add a little bit of same-GPU parallelism.
> 2. This wont' scale to anything more than small batches of 2-3 sites. And for larger datasets (i.e. more Z planes), the maximum allowed batch_size could even be one.

To me, this would suggest that parallelization is interesting, but needs to either be a user parameter (easy) or we need to be able to estimate the load (hard). I'd go for a user parameter to start with. Then we can collect some experience what works well and maybe come up with good heuristics for a given model or such.
Also, really good if this is something that can be varied in this task, because a user may e.g. decide to run cellpose at pyramid layer 1 (once this becomes possible using e.g. the ROI approach described above) and then could potentially run ~10 sites at once (given 10 Z layers) etc.

=> The flexibility is conceptually interesting, when and how we'll use it remains a bit of an open question though that we'll figure out more later I'd assume

**Comment by tcompa** [comment]
As of 60a27228f0fe3c2ebcf6ee0d1f317f36b9d798b6, the first image-labeling use case is roughly complete. It works for 2D or 3D images, always at the per-FOV level, and it has a `relabeling=True/False` argument. The labeling happens lazily over all FOV columns, and we can limit the number of concurrent cellpose executions with a `num_threads` argument. Relabeling is not done in a memory-efficient way, at the moment.

There are limits related to memory usage (different for labeling and relabeling) and runtimes (related to how many FOVs can be treated in parallel at the same time), which are better discussed in a new issue:
* https://github.com/fractal-analytics-platform/fractal/issues/108.

Issues related to the next use cases (whole-well segmentation in 2D and ROI-based scheme) are:
* https://github.com/fractal-analytics-platform/fractal/issues/104
* https://github.com/fractal-analytics-platform/fractal/issues/95

I'm closing this one.
